src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
+
+        n_theta = self.n_theta
+        if n_theta % 2 == 1:
+            logger.error('n_theta must be even')
+        n_theta = n_theta // 2
+        # The first two dimension of pf is of size n_rad x n_theta. Convert pf into an
+        # array of size (2xn_rad-1) x n_theta/2, that is, take entire ray
+        # through the origin, but take the angles only up PI. In fact, the original
+        # images are real, and thus each ray is conjugate symmetric. We therefore
+        # gain nothing by taking longer correlations (of length 2*n_rad-1 instead of n_rad),
+        # as the two halve are exactly the same. In real calculation of build_clmatrix below,
+        # We only taking shorter correlation and speed the computation by a factor of two.
+        self.pf = np.concatenate((np.flip(self.pf[1:, n_theta:], 0), self.pf[:, :n_theta]), 0)
+
+    def estimate_rotations(self):
+        """
+        Estimate orientation matrices for all 2D images
+
+        Subclasses should implement this function.
+        """
+        raise NotImplementedError('subclasses should implement this')
+
+    def save_rotations(self):
+        """
+        Save the estimated rotation matrices to new ImageSource object
+        """
+        # Generate a new ImageSource object and save estimated rotation matrices
+        # It will be simpler if it is allowed to modify the original one
+        src = ArrayImageSource(self.src.images(start=0, num=np.inf))
+        src.rots = self.rotations
+        return src
+
+    def build_clmatrix(self, n_check=None):
+        """
+        Build common-lines matrix from Fourier stack of 2D images
+
+        :param n_check: For each image/projection find its common-lines with
+            n_check images. If n_check is less than the total number of images,
+            a random subset of n_check images is used.
+        """
+
+        n_img = self.n_img
+        if n_check is None:
+            n_check = n_img
+        n_theta = self.n_theta
+        if n_theta % 2 == 1:
+            logger.error('n_theta must be even')
+        n_theta = n_theta // 2
+
+        # need to do a copy to prevent modifying self.pf for other functions
+        pf = self.pf.copy()
+
+        # Allocate local variables for return
+        # clstack represents the common lines matrix:
+        # clstack[i, j] and clstack[j ,i] contain the index
+        # of the common line of projections i and j. Namely,
+        # clstack[i,j] contains the index of the common line
+        # in the image i. clstack[j,i] contains the index of
+        # the common line in j image.
+        clstack = -np.ones((n_img, n_img))
+        # corrstack defines the correlation of the common line
+        # between image i and j. Since corrstack is symmetric,
+        # only above the diagonal entries are necessary.
+        # corrstack[i, j] measures how ''common'' is the between
+        # image i and j. Small value means high-similarity.
```

Comment:
The code use corrstack[i, j] = 0 to represent no need to check common line lines between i and j. This is in the opposite definition of correlation. It might be switched to 1 using the consistent definition.

Reply:
`corrstack[i, j]` is a dynamic variable changed with the shift step. I will keep the original definition to avoid conflicts.

Reply:
More comments are added.

Reply:
I don't understand. So it's not a correlation? Is it a distance? Let's use a more appropriate name in that case. What does it mean that there's no need to check the common lines? Does that mean that they're very similar?

Reply:
Changed the corrstack to skipstack and added more comments.  `skipstack[i, j]` measures how ''common'' is the between image i and j. Small value means high-similarity. We will use `skipstack[i, j] == 0` (including i=j) to represent that there is no need to check common line between i and j.

Reply:
So it's a distance (or dissimilarity, deviation, discrepancy) measure? Where does `skip` come from, in that case? Maybe just call it `cl_dist` in that case? At any rate, having zero indicate a special value here is confusing. How do we separate it from the case of two common lines being exactly identical (even if this is unlikely to happen)? A value of –1 might be better in this case. On the other hand, why store this data in the same array? Can't we have another one that stores a check flag separately from the distances?

Reply:
Renamed to `cl_dist` and initialize to -1 in commit fa2a5e8. It was removed from the class variables too.  The original author might think this matrix is quite large and it is not necessary have another one.

Reply:
The ` cl_dist ` is used for two purposes. The first one is used to identify the i, j pairs that no common-line calculation is needed (such as `j <= i` or some `j > i` in case of `n_check < n_img`). For this purpose `cl_dist[i, j]` is set to -1. It was set to zero in the original code but was changed to -1 based on Joakim's suggestion.  The other purpose is used to find the largest correlation (`sval`) when going through all 1D shifts and transferred to `1-sval` since we like zero to represent the case that i and j images are identical (similarity is more meaningful now).

Reply:
Okay. I see the one line where we use it for`sval > cl_dist[i, j]`, but we don't appear to use it again, yet we still modify it? Something doesn't make sense there.

Reply:
I still find this confusing, but I am not going to suggest any changes beyond the removal of the dead code that was made (which is actually the code that inverts the measure, but I digress...).

I am satisfied on this threat, but will leave open for Joakim to resolve.